src/devices/video/ef9345.cpp

```diff
@@ -229,26 +229,38 @@ void ef9345_device::set_busy_flag(int period)
 // draw a char in 40 char line mode
 void ef9345_device::draw_char_40(uint8_t *c, uint16_t x, uint16_t y)
 {
-	//verify size limit
-	if (y * 10 >= m_screen->height() || x * 8 >= m_screen->width())
-		return;
-
+	int scan_xsize,scan_ysize;
 	const rgb_t *palette = m_palette->palette()->entry_list_raw();
-	for(int i = 0; i < 10; i++)
-		for(int j = 0; j < 8; j++)
+
+	scan_xsize = m_screen->width() - (x * 8);
+	if ( scan_xsize > 8 )
+		scan_xsize = 8;
+
+	scan_ysize = m_screen->height() - (y * 10);
+	if ( scan_ysize > 10 )
+		scan_ysize = 10;
+
+	for(int i = 0; i < scan_ysize; i++)
+		for(int j = 0; j < scan_xsize; j++)
 				m_screen_out.pix32(y * 10 + i, x * 8 + j)  = palette[c[8 * i + j] & 0x07];
 }
 
 // draw a char in 80 char line mode
 void ef9345_device::draw_char_80(uint8_t *c, uint16_t x, uint16_t y)
 {
-	// verify size limit
-	if (y * 10 >= m_screen->height() || x * 6 >= m_screen->width())
-		return;
-
+	int scan_xsize,scan_ysize;
 	const rgb_t *palette = m_palette->palette()->entry_list_raw();
-	for(int i = 0; i < 10; i++)
-		for(int j = 0; j < 6; j++)
+
+	scan_xsize = m_screen->width() - (x * 6);
```

Comment:
Same here, make the locals `const` and use `std::min` to avoid the need for the `if` statements.